networkx/algorithms/community/louvain.py

```diff
@@ -0,0 +1,110 @@
+        then the algorithm stops and returns the resulting communities.
+
+    Returns
+    -------
+    list
+        A list of lists of sets. Each sublist of sets represents one
+        level of the tree and each set represents the communities with
+        all the nodes that constitute them.
+
+    References
+    ----------
+    .. [1] Blondel, V.D. et al. Fast unfolding of communities in
+       large networks. J. Stat. Mech 10008, 1-12(2008)
+    """
+
+    partitions = []
+    partition = [{u} for u in G.nodes()]
+    mod = modularity(G, partition)
+    current_graph = G.copy()
+
+    while True:
+        partition, improvement = _one_level(current_graph, deepcopy(partition), weight)
+        if not improvement:
+            break
+        new_mod = modularity(G, partition)
+        if new_mod - mod <= threshold:
+            break
+        partitions.append(partition)
+        current_graph = _gen_graph(G, partition)
+    return partitions
+
+
+def _one_level(G, partition, weight="weight"):
+    """Calculate one level of the tree"""
+    node2com = {u: i for i, u in enumerate(G.nodes())}
+    total_weights = {i: G.degree(u, weight=weight) for i, u in enumerate(G.nodes())}
+    degrees = G.degree(weight=weight)
+    rand_nodes = sample(G.nodes, len(G.nodes))
+    nb_moves = 1
+    m = G.size(weight=weight)
+    improvement = False
+    while nb_moves > 0:
+        nb_moves = 0
+        for u in rand_nodes:
+            nbrs = G.adj[u]
+            best_mod = 0
+            best_com = node2com[u]
+            partition[best_com].difference_update(G.nodes[u].get("graph", {u}))
+            weights2com = _neighbor_weights(u, nbrs, node2com, weight)
+            degree = degrees[u]
+            total_weights[best_com] -= degree
+            for nbr_com, weight in weights2com.items():
+                # Modularity gain
+                gain = weight - (total_weights[nbr_com] * degree) / m
+                if gain > best_mod:
+                    best_mod = gain
+                    best_com = nbr_com
+            partition[best_com].update(G.nodes[u].get("graph", {u}))
+            total_weights[best_com] += degree
+            if best_com != node2com[u]:
+                improvement = True
+                nb_moves += 1
+                node2com[u] = best_com
+    partition = list(filter(len, partition))
+    return partition, improvement
+
+
+def _neighbor_weights(node, nbrs, node2com, weight):
+    """Calculate node's neighbor communities and weights"""
+    weights = {}
+    for nbr, data in nbrs.items():
+        if nbr != node:
+            weights[node2com[nbr]] = weights.get(node2com[nbr], 0) + 2 * data.get(
+                weight, 1
+            )
+    return weights
+
+
+def _gen_graph(G, partition):
+    """Generate induced graph from given partition"""
+    H = nx.quotient_graph(G, partition, relabel=True)
+    for u in H.nodes:
+        if H.nodes[u]["nedges"] != 0:
+            H.add_edge(u, u, weight=H.nodes[u]["nedges"])
```

Comment:
Should the self_loop weight be “number of edges” between nodes in block “u”,  or “total weight of the edges”?
If it should be total weight, let’s give `quotient_graph` the input node_data(b) the stores b itself (instead of “graph”) and the total weight of edges from nodes in both nodes in b (instead of “nedges”). I think that will speed it up too.

Reply:
self_loops are the number of edges between nodes inside the same community only

Reply:
Hmmm… then maybe the “size” of the graph does change. We lose information about the weights of the edges within a community if we make the self-loop weight be the number of edges when we are using edge weights for other edges.

Reply:
Oh there was some misunderstanding, when I say number of edges I mean the weights between nodes of the same community. I interpreted wrongly what you said.

Reply:
OK... good -- then we should definitely define `node_data` to retain the original graph nodes in that partition as well as the sum of the edge weights. I think that will be faster than the current code anyway which stores the subgraph induced by the original nodes, the number of nodes, the number of edges and the density.  :}

Reply:
I tested with defining the `node_data` as you suggested but the timing was still pretty bad so I decided to write my own function to generate the new graph.

The new version takes about 0.5 seconds to generate a new graph of 27k nodes and 1500 communities.

Reply:
What was the timing before?  (using quotient_graph)

Reply:
About 200 seconds

Reply:
Wow!   Great job! :}  
That should speed things up a bit!  :}

Makes me want to revisit `quotient_graph`.  I'll open an issue about that.